### PR TITLE
AWS CloudWatch Output Plugin

### PR DESCRIPTION
**Pull request by skwong2**
I added an output plugin for sending metrics into AWS CloudWatch.

### CODE REVIEW

File: plugins/outputs/cloudwatch/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+## Amazon CloudWatch Output for Telegraf
+
+This plugin will send points to Amazon CloudWatch.
+
+## Amazon Authentication
+
+This plugin uses a credential chain for Authentication with the Kinesis API endpoint. In the following order the plugin
```
Comment (sparrc) [inline plugins/outputs/cloudwatch/README.md]: s/Kinesis/CloudWatch/g

File: plugins/outputs/cloudwatch/cloudwatch.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudwatch"
+
+	"github.com/influxdb/influxdb/client/v2"
+	"github.com/influxdb/telegraf/plugins/outputs"
+
+	"github.com/meirf/gopart"
+)
+
+type CloudWatchOutput struct {
```
Comment (sparrc) [inline plugins/outputs/cloudwatch/cloudwatch.go]: rename struct to just `CloudWatch`

File: plugins/outputs/cloudwatch/cloudwatch.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+		log.Printf("cloudwatch: Error in ListMetrics API call : %+v \n", err.Error())
+	}
+
+	c.svc = svc
+
+	return err
+}
+
+func (c *CloudWatchOutput) Close() error {
+	return errors.New("Error")
```
Comment (sparrc) [inline plugins/outputs/cloudwatch/cloudwatch.go]: This should just `return nil`, I'm going to fix the kinesis output to do the same

File: plugins/outputs/cloudwatch/cloudwatch.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+		}
+
+		i += 1
+	}
+
+	return datums
+}
+
+// Make a list of Dimensions by using a Point's tags.
+func buildDimensions(ptTags map[string]string) []*cloudwatch.Dimension {
```
Comment (skwong2) [inline plugins/outputs/cloudwatch/cloudwatch.go]: There's actually a bug here.  CloudWatch only supports up to 10 dimensions per datum.  Some points have more than 10 tags.  I'll fix this but I'm not sure the best approach here.  I'm thinking of sorting and keeping the first 10.

Comment (sparrc) [inline plugins/outputs/cloudwatch/cloudwatch.go]: hmm your solution seems reasonable enough, I'm not sure the best way to get around that either, some tags will have to be dropped

File: plugins/outputs/cloudwatch/cloudwatch.go [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+// Write data for a single point. A point can have many fields and one field
+// is equal to one MetricDatum. There is a limit on how many MetricDatums a
+// request can have so we process one Point at a time.
+func (c *CloudWatch) WriteSinglePoint(point *client.Point) error {
+	datums := buildMetricDatum(point)
+
+	const maxDatumsPerCall = 20 // PutMetricData only supports up to 20 data points per call
+
+	for idxRange := range gopart.Partition(len(datums), maxDatumsPerCall) {
```
Comment (sparrc) [inline plugins/outputs/cloudwatch/cloudwatch.go]: I agree that using `gopart` here makes it elegant, but I don't think it justifies adding another 3rd-party dependency to Telegraf. Could you add a little extra code and do this yourself? One option that comes to mind would be to have `buildMetricDatum` do the partitioning itself and return `[][]*cloudwatch.MetricDatum`

**Comment by sparrc** [comment]
this looks great, thanks @skwong2!

please address my comment and write some unit tests and it will be ready to merge :+1:

File: plugins/outputs/cloudwatch/cloudwatch.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+	if err != nil {
+		log.Printf("CloudWatch: Unable to write to CloudWatch : %+v \n", err.Error())
+	}
+
+	return err
+}
+
+// Make a MetricDatum for each field in a Point. Only fields with values that can be
+// converted to float64 are supported. Non-supported fields are skipped.
+func buildMetricDatum(point *client.Point) []*cloudwatch.MetricDatum {
```
Comment (sparrc) [inline plugins/outputs/cloudwatch/cloudwatch.go]: this function can be unit tested

File: plugins/outputs/cloudwatch/cloudwatch.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+		i += 1
+	}
+
+	return datums
+}
+
+// Make a list of Dimensions by using a Point's tags. CloudWatch supports up to
+// 10 dimensions per metric so we only keep up to the first 10 alphabetically.
+// This always includes the "host" tag if it exists.
+func buildDimensions(ptTags map[string]string) []*cloudwatch.Dimension {
```
Comment (sparrc) [inline plugins/outputs/cloudwatch/cloudwatch.go]: this function can be unit tested

**Comment by skwong2** [comment]
@sparrc  Not sure why the tests for other plugins failed there.  Anyway I can kick off the build process again?

**Comment by sparrc** [comment]
@skwong2 I'll kick it

**Comment by sparrc** [comment]
thanks a bunch @skwong2!

**Comment by skwong2** [comment]
@sparrc Thanks! Does Telegraf have a nightly build? Or is there a date for the next release?

**Comment by sparrc** [comment]
No nightly builds yet :neutral_face: I'll release 0.10.1 on Tuesday

**Pull request closed without merging**